indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java

```diff
@@ -0,0 +1,159 @@
+  private static final Logger logger = Logger.getLogger(IncludeExcludeFilter.class.getName());
+  public static final String FILTER_CONFIG_PREFIX = "includeExcludeFilter.";
+  public static final String INCLUDE_RULE_PREFIX = FILTER_CONFIG_PREFIX + "include.regex.";
+  public static final String EXCLUDE_RULE_PREFIX = FILTER_CONFIG_PREFIX + "exclude.regex.";
+
+  @VisibleForTesting final List<Rule<String>> includeRules;
+  @VisibleForTesting final List<Rule<String>> excludeRules;
+
+  public IncludeExcludeFilter(List<Rule<String>> includeRules, List<Rule<String>> excludeRules) {
+    this.includeRules = includeRules;
+    this.excludeRules = excludeRules;
+  }
+
+  public static IncludeExcludeFilter fromConfiguration() {
+    checkState(Configuration.isInitialized());
+    Set<String> filterProperties = Configuration.getConfig()
+        .stringPropertyNames()
+        .stream()
+        .filter(key -> key.startsWith(FILTER_CONFIG_PREFIX))
+        .collect(Collectors.toSet());
+    if (filterProperties.isEmpty()) {
+      return new IncludeExcludeFilter(ImmutableList.of(), ImmutableList.of());
+    }
+    List<Rule<String>> includeRules = new ArrayList<IncludeExcludeFilter.Rule<String>>();
+    List<Rule<String>> excludeRules = new ArrayList<IncludeExcludeFilter.Rule<String>>();
+    for (String propertyName : filterProperties) {
+      String propertyValue = Configuration.getString(propertyName, null).get();
+      logger.log(Level.FINEST, "Processing include/exclude rule {0}: {1}",
+          new Object[] { propertyName, propertyValue });
+      try {
+        if (propertyName.startsWith(INCLUDE_RULE_PREFIX)) {
+          includeRules.add(new Rule<>(new RegexPredicate(propertyValue)));
+        } else if (propertyName.startsWith(EXCLUDE_RULE_PREFIX)) {
+          excludeRules.add(new Rule<>(new RegexPredicate(propertyValue)));
+        }
+      } catch (PatternSyntaxException e) {
+        throw new InvalidConfigurationException("Invalid regex pattern " + propertyValue, e);
+      }
+    }
+    return new IncludeExcludeFilter(includeRules, excludeRules);
+  }
+
+  /**
+   * Returns true if the given string is included based on the configured include/exclude
+   * patterns.
+   *
+   * @param value a value to test
+   * @return true if the value is included based on the configuration
+   */
+  public boolean isAllowed(String value) {
+    boolean exclude = evaluateRules(excludeRules, value, false /* nothing is excluded */);
+    if (exclude) {
+      logger.log(Level.FINEST, "excluding " + value);
+      return false;
+    }
+    return evaluateRules(includeRules, value, true /* everything is included */);
+  }
+
+  private boolean evaluateRules(List<Rule<String>> rules, String value, boolean emptyRulesOutcome) {
+    if (rules.isEmpty()) {
+      return emptyRulesOutcome;
+    }
+    return rules.stream().map(r -> r.eval(value)).filter(e -> e == true).findFirst().orElse(false);
+  }
+
+  private static class Rule<T> {
```

Comment:
This doesn't seem to be carrying much weight. Why not just use Predicate directly?